# Seurat-RNA-Seq-analysis-ScRNA data is generated using the 10X genomics platform. The cellranger pipeline outputs two types of feature-barcode matrices. Each element of the matrix is the number of UMIs (Short random molecular tags which are added to DNA fragments in library preparation process before PCR amplification. UMIs are used to identify input DNA molecules) associated with a feature (row) and a barcode (column).
Filtered feature-barcode matrix Contains only detected cellular barcodes. For Targeted Gene Expression samples, non-targeted genes are removed from the filtered matrix.
Each matrix is stored in the Market Exchange Format (MEX) for sparse matrices. It also contains gzipped TSV files with feature and barcode sequences corresponding to row and column indices respectively.
Start by reading in the data. The Read10X() function reads in the output of the cellranger pipeline from 10X, returning a unique molecular identified (UMI) count matrix. The values in this matrix represent the number of molecules for each feature (i.e. gene; row) that are detected in each cell (column).
Next use the count matrix to create a Seurat object. The object serves as a container that contains both data (like the count matrix) and analysis (like PCA, or clustering results) for a single-cell dataset.
scRNA-seq raw data includes reads with cell barcode (unique nucleic acid sequences, termed barcodes, which are used to label individual cells, so that they can be tracked through space and time in scRNA-seq) and UMIs. Before alignment of reads to genome, reads can be grouped using cell barcodes and the frequency of each read is estimated per cell per gene using UMIs. After alignment and frequency calculations, we have a gene expression table containing cells represented in the columns and genes represented in the rows. This is what is analyzed in scRNA-seq comlutational workflow.
Here is an image of ScRNA seq workflow- ![image](https://user-images.githubusercontent.com/114133070/198524323-3d33551d-b628-439f-a494-0f3a89c9587e.png)

#
